storage/ndb/rest-server/rest-api-server/internal/security/apikey/apikey_cache/hw_apikey_cache/hw_api_cache.go

```
@@ -0,0 +1,319 @@
+	hwc.key2UserDBsCacheLock.Lock()
+	defer hwc.key2UserDBsCacheLock.Unlock()
+
+	if log.IsInfo() {
+		log.Info("Shutting down API Key Cache")
+	}
+
+	for _, udbs := range hwc.key2UserDBsCache {
+		udbs.ticker.Stop()
+		udbs.evicted = true
+	}
+
+	hwc.key2UserDBsCache = make(map[string]*UserDBs)
+
+	return nil
+}
+
+// update the cache entry by fetching the API key from backend
+func (hwc *HWAPIKeyCache) UpdateCache(apiKey *string) error {
+
+	// if the entry does not already exist in the
+	// cache then multiple clients will try to read and
+	// update the API key from the backend simultaneously.
+	// Trying to prevent multiple writers here
+
+	// first check using read lock
+	hwc.key2UserDBsCacheLock.RLock()
+	_, ok := hwc.key2UserDBsCache[*apiKey]
+	hwc.key2UserDBsCacheLock.RUnlock()
+
+	if !ok {
+		// Continue with write lock
+		hwc.key2UserDBsCacheLock.Lock()
+
+		udbs, ok := hwc.key2UserDBsCache[*apiKey]
+		if !ok { // the entry still does not exists. insert a new row
+			udbs = &UserDBs{}
+			udbs.refreshInterval = hwc.refreshIntervalWithJitter()
+			hwc.key2UserDBsCache[*apiKey] = udbs
+			hwc.startUpdateTicker(apiKey, udbs)
+		}
+		hwc.key2UserDBsCacheLock.Unlock()
+	}
+
+	// the entry already exists. Ticker will update it
+	return nil
+}
+
+func (hwc *HWAPIKeyCache) startUpdateTicker(apiKey *string, udbs *UserDBs) error {
+	started := false
+	go hwc.cacheEntryUpdater(apiKey, &started)
+
+	//wait for go routine to start
+	for {
+		if started {
+			if log.IsDebug() {
+				log.Debugf("API Key cache updater is started for %s. Refresh Interval: %v ", *apiKey, udbs.refreshInterval)
+			}
+			return nil
+		} else {
+			time.Sleep(50 * time.Microsecond)
+		}
+	}
+}
+
+func (hwc *HWAPIKeyCache) cacheEntryUpdater(apiKey *string, started *bool) {
+
+	udbs, ok := hwc.key2UserDBsCache[*apiKey] // not need for read lock here as the caller holds write lock
+	if !ok {
+		log.Errorf("Cache updater failed. Report programming error. API Key %s", *apiKey)
+		return
```

Why not return an error and let the caller decide what to do?

---

This is async job to update the cache. It can only log the errors

---

But then let the worker decide that. You don't even have to do 'err :=' if you don't want to.